init/modules/EESSI/2023.06.lua

```diff
@@ -17,56 +17,120 @@ local eessi_os_type = "linux"
-            -- see https://gitlab.com/eessi/support/-/issues/37 
-            if archdetect_filter_cpu == "x86_64/amd/zen4" then
-                archdetect_filter_cpu = "x86_64/amd/zen3"
-                if mode() == "load" then
-                    LmodMessage("Sticking to " .. archdetect_filter_cpu .. " for now, since optimized installations for AMD Genoa (Zen4) are a work in progress.")
+    -- EESSI_ARCHDETECT_OPTIONS is set by the script (_if_ it was called)
+    archdetect_options = os.getenv("EESSI_ARCHDETECT_OPTIONS") or archdetect_options
+    if archdetect_options then
+        eessiDebug("Got archdetect CPU options: " .. archdetect_options)
+        for archdetect_filter_cpu in string.gmatch(archdetect_options, "([^" .. ":" .. "]+)") do
+            if isDir(pathJoin(eessi_prefix, "software", eessi_os_type, archdetect_filter_cpu, "software")) then
+                -- use x86_64/amd/zen3 for now when AMD Genoa (Zen4) CPU is detected,
+                -- since optimized software installations for Zen4 are a work-in-progress,
+                -- see https://gitlab.com/eessi/support/-/issues/37
+                if (archdetect_filter_cpu == "x86_64/amd/zen4" and not os.getenv("EESSI_SOFTWARE_SUBDIR_OVERRIDE") == "x86_64/amd/zen4") then
+                    archdetect_filter_cpu = "x86_64/amd/zen3"
+                    if mode() == "load" then
+                        LmodMessage("Sticking to " .. archdetect_filter_cpu .. " for now, since optimized installations for AMD Genoa (Zen4) are a work in progress.")
+                    end
                 end
+                eessiDebug("Selected archdetect CPU: " .. archdetect_filter_cpu)
+                return archdetect_filter_cpu
             end
-            return archdetect_filter_cpu
         end
+        LmodError("Software directory check for the detected architecture failed")
+    else
+        -- Still need to return something
+        return nil
+    end
+end
+function archdetect_accel()
+    local script = pathJoin(eessi_prefix, 'init', 'lmod_eessi_archdetect_wrapper_accel.sh')
+    -- for unload mode, we need to grab the value before it is unset
+    local archdetect_accel = os.getenv("EESSI_ACCEL_SUBDIR") or (os.getenv("EESSI_ACCELERATOR_TARGET_OVERRIDE") or "")
+    if not os.getenv("EESSI_ACCELERATOR_TARGET_OVERRIDE ") then
+        if convertToCanonical(LmodVersion()) < convertToCanonical("8.6") then
+            LmodError("Loading this modulefile requires using Lmod version >= 8.6, but you can export EESSI_ACCELERATOR_TARGET_OVERRIDE to the available accelerator architecture in the form of: accel/nvidia/cc80")
+        end
+        source_sh("bash", script)
     end
-    LmodError("Software directory check for the detected architecture failed")
+    archdetect_accel = os.getenv("EESSI_ACCEL_SUBDIR") or archdetect_accel
+    eessiDebug("Got archdetect accel option: " .. archdetect_accel)
+    return archdetect_accel
 end
 local archdetect = archdetect_cpu()
```

Comment:
I'm wondering if we should add comments with examples to each of these. I think our future selves / others maintaining this script could find that helpful. Especially lower down, at 

```
eessi_module_path_accel = pathJoin(eessi_accel_software_path, archdetect_accel, eessi_modules_subdir)
```

I was a bit puzzled about an `archdetect_accel` being appended to an `eessi_accel_software_path`. I was thinking _both_ would contain something like ccXX, but that's not true (or so I think): I guess `eessi_accel_software_path` is just the CPU-based prefix (to which we then append the accelerator part of the prefix).

While it's possible to figure it out (or have things printed with the nice new debug option), I think adding an example to each would facilitate future code maintenance.

Reply:
Done in various commits